ReactAndroid/src/main/java/com/facebook/react/uimanager/BaseViewManager.java

```diff
@@ -125,32 +138,53 @@ public void setAccessibilityComponentType(@Nonnull T view, String accessibilityC
   @ReactProp(name = PROP_ACCESSIBILITY_HINT)
   public void setAccessibilityHint(@Nonnull T view, String accessibilityHint) {
     view.setTag(R.id.accessibility_hint, accessibilityHint);
+    updateViewContentDescription(view);
   }
 
   @ReactProp(name = PROP_ACCESSIBILITY_ROLE)
   public void setAccessibilityRole(@Nonnull T view, @Nullable String accessibilityRole) {
     if (accessibilityRole == null) {
       return;
     }
-
     view.setTag(R.id.accessibility_role, AccessibilityRole.fromValue(accessibilityRole));
   }
 
   @ReactProp(name = PROP_ACCESSIBILITY_STATES)
   public void setViewStates(@Nonnull T view, @Nullable ReadableArray accessibilityStates) {
-    view.setSelected(false);
-    view.setEnabled(true);
     if (accessibilityStates == null) {
       return;
     }
+    view.setTag(R.id.accessibility_states, accessibilityStates);
     for (int i = 0; i < accessibilityStates.size(); i++) {
       String state = accessibilityStates.getString(i);
-      if (state.equals("selected")) {
-        view.setSelected(true);
-      } else if (state.equals("disabled")) {
-        view.setEnabled(false);
+      if (sStateDescription.containsKey(state)) {
+        updateViewContentDescription(view);
+      }
+    }
+  }
+
+  private void updateViewContentDescription(@Nonnull T view) {
```

Comment:
Background: Chris McMeeking, mobile accessibility expert, member WCAG Mobile Accessibility Task Force, etc. etc.

I love everything about this pull request except this function. Appending things to the content description is icky. Akin to Off Screen text in the web world. I'd rather see a feature request sent to Google. Programmatic association with Roles and States is crazy important. If TalkBack has not done their homework on how to communicate and keep track of those states, that is TalkBacks problem.

A solution like this will fragment the Android Accessibility Ecosystem even further than it already is. IMHO an effort like this should stick within the Operating Systems capabilities, and if those capabilities lead us to the conclusion that the platform is not Accessible, we should unitedly and vocally raise a fuss about this. But, appending a bunch of crud into Off Screen text and pretend like that fixes things, is a bad idea. We'd be better to embrace Android's features and refrain from fragmenting the Operating System even more than it already is.

Reply:
I respectfully disagree. We're dealing with more than simply Android support with this pull request-- we're thinking about iOS and Web as well. And we can't afford to "dumb down" a toolkit like React Native to the least common denominator. If we do that, then we have no checkbox role, no checked state, etc. since iOS doesn't natively support those.

I wholeheartedly agree that this solution doesn't work well in braille, or for single switch access. But IMHO, improvements in speech access and support are better than no improvements at all.

The way this is architected, Javascript components expose the states and roles semantically, so if someday the situation on Android or iOS changes, we can re-implement the platform support layer to use native roles and states, which I think is the right approach.

Reply:
You can have roles and traits that you don't communicate. Off-Screen text is a known non-solution in the Web Sphere, and that is all that this is.

Having a trait in react native, and admitting that the Android operating system doesn't respect it, seems perfectly reasonable to me.

Reply:
> But IMHO, improvements in speech access and support are better than no improvements at all.

This is assuming that speaking roles is valuable. It frequently isn't. It is the other, programmatic behavioral responses to traits that matter much more.

Reply:
Quote from Android Accessibility Best Practices Page: Many accessibility services, such as TalkBack and BrailleBack, automatically announce an element's type after announcing its label, so you shouldn't include element types in your labels. For example, "submit" is a good label for a Button object, but "submitButton" isn't a good label.

If Android's best practices are wrong, we should ask them to change them. But a platform like React Native deciding that those best practices are wrong... bad plan.

Reply:
Guess we'll agree to disagree on this. We are simply extending work Facebook began last summer, and I haven't seen from you an alternate  proposal which solves the problem(s) this PR addresses.

Reply:
Let me rephrase this for the TalkBack centric developer. Try this out:

1. Turn on TalkBack
2. Change it to `Controls` Navigation mode.
    1. Swipe Up until it says "Controls"
3. Note that most of your Pseudo Android roles are not marked up properly.

It is the ability to easily explore controls amidst a miriad of informative elements that makes a control a control NOT what it is announced as. Most of your pseudo roles are controls, and are lacking this behavior, which is significantly more important than appending that information to the "speakable text".

Reply:
> Given this lack of clear purpose, such a role would be used very arbitrarily across the react-native ecosystem

The React Native ecosystem is not limited to mobile and includes desktop apps

Reply:
@chriscm2006, I think there may be some confusion here. This method only adds concatenated text for hints (which Android only supports if they are associated with an action, not just a generic string like iOS or web), and 5 different states (defined in the sStateDescription map). These states (on, off, expanded, collapsed, and busy) are not ones that allow users to change navigation modes to easily find.  Roles and States that can be semantically expressed in Android are handled correctly in the accessibility delegate here:

https://github.com/facebook/react-native/pull/24095/files/35a93f4f995bc849d90af0e8dd06aac873f66aff#diff-eee51b170ff20e1fc3aa718ca2d17bb1R116

I understand your concern though, as this definitely isn't a best practice approach when it can be avoided.  But if the choice was between not telling a user something is expanded (which is usually obvious visually), versus the user not having control to disable this announcement (as they can do by disabling state information), I'd err on the side of providing too much information over not enough. While this could be annoying for power users, the vast majority of users are not adjusting their settings from the default, and would simply be missing this state information which could be important to the understanding of the UI.

All that being said, I do think that on/off could be removed from this list, as semantically it used in the same way as "checked" and "unchecked".  In fact, when Talkback sees an element with the "Switch" role (the only one I know to use "on" and "off" as its labels), it uses the value of the checked property in the AccessibilityNodeInfo to append "on" or "off" to its readback.  Logic in the talkback compositor here:

https://github.com/google/talkback/blob/e69d4731fce02bb9e69613d0e48c29033cad4a98/compositor/src/main/res/raw/compositor.json#L998

If we want to keep the ReactNative API to use the state of "on" and "off", we can do so, but we should map this to setting the checkable and checked properties in the AccessibilityDelegate rather than appending them to the content description here.

Reply:
I think if I saw some of the items handled in the semantic way when it can be, I would feel better. It's one thing to create things that aren't there... I'll go quiet on that point, but it's another to ignore Operating System support that is there. The two examples that I can't get over.

- CollectionItemInfo for menu items.
- Ensuring that everything that is a control is either "mocked" as such, or throws exceptions or warnings when is isn't used properly.

Reply:
> > Given this lack of clear purpose, such a role would be used very arbitrarily across the react-native ecosystem
> 
> The React Native ecosystem is not limited to mobile and includes desktop apps

I'm not asking to drop those roles, I'm asking to ignore or mark up those roles properly for Android. Not all Reac Native apps need to behave the same way. In fact they shouldn't. They should behave the way the platform would normally behave.

Reply:
To flip this argument around-- iOS doesn't have a checkbox trait or a checked trait. It's essential to know that something is a checkbox, and whether it is checked or unchedked. Android supports both a checkbox role convention (className == "android.widget.CheckBox" and the isChecked() method). So does that mean we shouldn't surface the checkbox role and associated checked/unchecked states on iOS because it doesn't support them?

We have used Android conventions where they exist to annotate the roles and states they represent, so setting class name as you suggest, as well as setting node metadata such as isChecked().

Happy to remove on/off in favor of checked/unchedked.

You can reasonably argue that menu and menuitem are not useful as roles on mobile, but that's a different argument than saying the way in which we support them is wrong. Using CollectionInfo and CollectionItemInfo only surfaces to the user that this is a list/grid, not that it's a menu. Knowing whether something is a menu or a standard list *could* be important IMHO.

We're using the roleDescription paradigm to annotate roles that don't have Android native analogs (equivalent Java class names). This is a paradigm introduced by Google for Chromium accessibility on Android where everything has a class name of "android.view.View".

Only states without AccessibilityNodeInfo analogs are concatenated to the content description, where we deemed such state information was useful.

I'm in favor of using Android-specific markup where it exists. If we've missed some standard markup, let us know and we'll fix it.

In summary:

* roles never get appended to the content description, they either cause the AccessibilityNodeInfo's class name to be set to the closest native analog, or a string description of the role to be placed in the roleDescription extra.
* States set the AccessibilityNodeInfo property where it exists, otherwise a string description of the state is appended to the content description.

Reply:
Checked/unchecked in favor of on/off I think is a good idea. Though you could also look into:

`setTextOn(), setTextOff()`

For scenarios when you want to be more expressive about the state of a given toggle-able trait.

I'd have to look closer at the roles where you add states, but the best Android Markup for times when you need to convey State AND Name, is to associate the stateful control and a `TextView` with a `labelFor` attribute. This may or may not be a solution in this case, I feel like that may turn this effort into something larger than it is...